distribution/tools/launchers/src/test/java/org/elasticsearch/tools/launchers/JvmOptionsParserTests.java

```diff
@@ -149,6 +156,108 @@ public void testComplexOptions() throws IOException {
+            StandardOpenOption.APPEND
+        );
+        Files.write(
+            config.resolve("jvm.options.d").resolve("heap.options"),
+            List.of("# comment", "-Xms384m", "-Xmx384m"),
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.APPEND
+        );
+        final JvmOptionsParser parser = new JvmOptionsParser();
+        final List<String> jvmOptions = parser.readJvmOptionsFiles(config);
+        assertThat(jvmOptions, contains("-Xms256m", "-Xmx256m", "-Xms384m", "-Xmx384m"));
+    }
+
+    public void testReadJvmOptionsDirectoryInOrder() throws IOException, JvmOptionsParser.JvmOptionsFileParserException {
+        final Path config = newTempDir();
+        Files.createDirectory(config.resolve("jvm.options.d"));
+        Files.write(
+            config.resolve("jvm.options"),
+            List.of("# comment", "-Xms256m", "-Xmx256m"),
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.APPEND
+        );
+        Files.write(
+            config.resolve("jvm.options.d").resolve("first.options"),
+            List.of("# comment", "-Xms384m", "-Xmx384m"),
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.APPEND
+        );
+        Files.write(
+            config.resolve("jvm.options.d").resolve("second.options"),
+            List.of("# comment", "-Xms512m", "-Xmx512m"),
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.APPEND
+        );
+        final JvmOptionsParser parser = new JvmOptionsParser();
+        final List<String> jvmOptions = parser.readJvmOptionsFiles(config);
+        assertThat(jvmOptions, contains("-Xms256m", "-Xmx256m", "-Xms384m", "-Xmx384m", "-Xms512m", "-Xmx512m"));
+    }
+
+    public void testReadJvmOptionsDirectoryIgnoresFilesNotNamedOptions() throws IOException,
+        JvmOptionsParser.JvmOptionsFileParserException {
+        final Path config = newTempDir();
+        Files.createFile(config.resolve("jvm.options"));
+        Files.createDirectory(config.resolve("jvm.options.d"));
+        Files.write(
+            config.resolve("jvm.options.d").resolve("heap.not-named-options"),
+            List.of("# comment", "-Xms256m", "-Xmx256m"),
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.APPEND
+        );
+        final JvmOptionsParser parser = new JvmOptionsParser();
+        final List<String> jvmOptions = parser.readJvmOptionsFiles(config);
+        assertThat(jvmOptions, empty());
+    }
+
+    public void testFileContainsInvalidLinesThrowsParserException() throws IOException {
+        final Path config = newTempDir();
+        final Path rootJvmOptions = config.resolve("jvm.options");
+        Files.write(rootJvmOptions, List.of("XX:+UseG1GC"), StandardOpenOption.CREATE_NEW, StandardOpenOption.APPEND);
+        try {
+            final JvmOptionsParser parser = new JvmOptionsParser();
+            parser.readJvmOptionsFiles(config);
+            fail();
```

Comment:
Same as above, seems it would be helpful to provide a message here to avoid a generic `AssertionError`.